symbolic-base/src/ZkFold/Symbolic/Compiler/ArithmeticCircuit.hs

```diff
@@ -33,46 +33,115 @@ module ZkFold.Symbolic.Compiler.ArithmeticCircuit (
+import           ZkFold.Base.Data.Product                                (fstP, sndP)
+import           ZkFold.Prelude                                          (length)
+import           ZkFold.Symbolic.Class                                   (fromCircuit2F)
+import           ZkFold.Symbolic.Compiler.ArithmeticCircuit.Instance     ()
+import           ZkFold.Symbolic.Compiler.ArithmeticCircuit.Internal     (Arithmetic, ArithmeticCircuit (..),
+                                                                          Constraint, SysVar (..), Var (..),
+                                                                          WitVar (..), acInput, eval, eval1, exec,
+                                                                          exec1, hlmap, witnessGenerator)
 import           ZkFold.Symbolic.Compiler.ArithmeticCircuit.Map
-import           ZkFold.Symbolic.Data.Combinators                    (expansion)
-import           ZkFold.Symbolic.MonadCircuit                        (MonadCircuit (..))
+import           ZkFold.Symbolic.Compiler.ArithmeticCircuit.Witness      (WitnessF (..))
+import           ZkFold.Symbolic.Data.Combinators                        (expansion)
+import           ZkFold.Symbolic.MonadCircuit                            (MonadCircuit (..))
 
 --------------------------------- High-level functions --------------------------------
 
 -- | Optimizes the constraint system.
 --
--- TODO: Implement nontrivial optimizations.
-optimize :: ArithmeticCircuit a p i o -> ArithmeticCircuit a p i o
-optimize = id
+optimize1 :: forall a p i o. ( Arithmetic a, Ord (Rep i))
+  => SysVar i -> a -> State (ArithmeticCircuit a p i o) ()
+optimize1 v k = case v of
+  NewVar nv -> modify (\(f :: ArithmeticCircuit a p i o) ->
+    f {
+      acSystem = M.map optPoly $ acSystem f,
+      acRange = optRanges $ acRange f,
+      acWitness = M.map optWitVar $ M.delete nv $ acWitness f
+      })
+  _ -> error "This shouldn't happen"
+  where
+    optMono :: (a, Mono (SysVar i) Natural) -> (a, Mono (SysVar i) Natural)
+    optMono mono@(c, M m) =
+      case M.lookup v m of
+        Just y -> (c * k ^ y, M $ delete v m)
+        _      -> mono
+
+    optPoly :: Poly a (SysVar i) Natural -> Poly a (SysVar i) Natural
+    optPoly (P (p :: [(a, Mono (SysVar i) Natural)])) = P $
+      map optMono p
+
+    optRanges :: MM.MonoidalMap a (S.Set (SysVar i)) -> MM.MonoidalMap a (S.Set (SysVar i))
+    optRanges = MM.mapWithKey (\i s ->
+      if S.member v s && k <= i
+        then S.filter (/= v) s
+        else s)
+
+    optWitVar ::  WitnessF a (WitVar p i) -> WitnessF a (WitVar p i)
+    optWitVar (WitnessF w)  = WitnessF $ \ww -> w $ \case
+      WSysVar nV@(NewVar _) ->
+        if v == nV
+        then ww (WSysVar nV)
+        else fromConstant k
+      x -> ww x
+
+optimize ::
+  ( Arithmetic a, Ord (Rep i)) =>
+  ArithmeticCircuit a p i o  -> ArithmeticCircuit a p i o
+optimize = until (not . haveOptimal) f
```

Comment:
I think we can avoid a `haveOptimal` call by matching on the emptiness of `[toConstVar p | ...]`